### DISCUSSION TITLE
Conditional requester-pays

### DISCUSSION THREAD

**Discussion by moradology** [developmentseed/titiler#758] [Q&A] [answered]
I'm trying to think about how the possibility of *conditional* requester pays reads might be handled in TiTIler and am wondering if it is currently possible. Here's an example STAC asset with a 'requester-pays' field that ideally could be picked up and passed through to the `TilerFactory` that ultimately does the relevant reads:
<img width="983" alt="image" src="https://github.com/developmentseed/titiler/assets/1977405/761ef7ef-b50f-4933-aed9-2b789e81960c">

One option would seem to be a custom `TilerFactory` that constructs the relevant environment on the basis of metadata invoked during `Reader` instantiation. Perhaps as simple as adding a reference to `self.reader(src_path, **reader_params).requester_pays` (or something like it) prior to these lines for inclusion in the `rasterio.Env`: https://github.com/developmentseed/titiler/blob/main/src/titiler/core/titiler/core/factory.py#L416-L417

Perhaps there's something I've missed? Are there any other ideas?

**Comment by vincentsarago** [comment, marked as answer]
In theory you could do something like: 
https://github.com/cogeotiff/rio-tiler/blob/874f1410e04c3e46bd50f6a272340787ef39fb55/rio_tiler/io/stac.py#L302-L303

**Reply by moradology** [reply]
This looks very clean but I'm a bit confused about where the `env` is surfaced for GDAL to use. I'd expect, perhaps, to see something here in the `read` block, but `env` is nowhere to be found (https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/reader.py). Perhaps you can share your wisdom on how this happens?

**Reply by vincentsarago** [reply]
this is then passed here https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/io/base.py#L501
